integration_tests/lock_test.go

```diff
@@ -1248,3 +1254,269 @@ func (s *testLockWithTiKVSuite) TestPrewriteCheckForUpdateTS() {
+			lockCtx := kv.NewLockCtx(t1.StartTS(), 200, time.Now())
+			s.NoError(t1.LockKeys(ctx, lockCtx, pk))
+			s.NoError(t1.Set(pk, v1))
+
+			simulatedCommiter, err := t1.NewCommitter(1)
+			s.NoError(err)
+
+			t2, err := s.store.Begin()
+			s.NoError(err)
+			t2.SetEnableAsyncCommit(true)
+			s.NoError(t2.Set(key1, v2))
+			s.NoError(t2.Set(key2, v2))
+
+			t1.StartAggressiveLocking()
+			lockTime := time.Now()
+			lockCtx = kv.NewLockCtx(t1.StartTS(), 200, lockTime)
+			s.NoError(t1.LockKeys(ctx, lockCtx, key1))
+			t1.CancelAggressiveLocking(ctx)
+			// Wait the pessimistic lock of t1 on key1 to be rolled back.
+			s.checkIsKeyLocked(key1, false)
+
+			// Try to make transaction t1 and t2 commit at the same commitTS.
+			// First, block at finishing getting minCommitTS from PD and before prewriting
+			s.NoError(failpoint.Enable("tikvclient/beforePrewrite", "pause"))
+			resCh := make(chan error)
+			go func() {
+				resCh <- t1.Commit(ctx)
+			}()
+			go func() {
+				resCh <- t2.Commit(ctx)
+			}()
+			select {
+			case <-resCh:
+				s.Fail("commit not blocked")
+			case <-time.After(time.Millisecond * 100):
+			}
+
+			// Send a read request to TiKV to update max_ts
+			currentTS, err := s.store.CurrentTimestamp(oracle.GlobalTxnScope)
+			s.NoError(err)
+			snapshot := s.store.GetSnapshot(currentTS)
+			_, err = snapshot.Get(context.Background(), key2)
+			s.Equal(tikverr.ErrNotExist, err)
+
+			// Continue committing
+			s.NoError(failpoint.Disable("tikvclient/beforePrewrite"))
+			s.NoError(<-resCh)
+			s.NoError(<-resCh)
+
+			if t1.GetCommitTS() != t2.GetCommitTS() {
+				return false
+			}
+
+			checkDataConsistency := func() {
+				currentTS, err = s.store.CurrentTimestamp(oracle.GlobalTxnScope)
+				s.NoError(err)
+				snapshot = s.store.GetSnapshot(currentTS)
+				v, err := snapshot.Get(context.Background(), pk)
+				s.NoError(err)
+				s.Equal(v1, v)
+				v, err = snapshot.Get(context.Background(), key2)
+				s.NoError(err)
+				s.Equal(v2, v)
+				v, err = snapshot.Get(context.Background(), key1)
+				s.NoError(err)
+				s.Equal(v2, v)
+			}
+
+			checkDataConsistency()
+
+			// Simulate a repeated stale request of t1 locking key1 sent to TiKV
+			{
+				simulatedCommiter.SetPrimaryKey(pk)
+				simulatedCommiter.SetForUpdateTS(t1.StartTS())
+				lockCtx := kv.NewLockCtx(t1.StartTS(), 200, lockTime)
+				mutations := transaction.NewPlainMutations(1)
+				mutations.Push(kvrpcpb.Op_PessimisticLock, key1, nil, true, false, false, false)
+				err = simulatedCommiter.PessimisticLockMutations(ctx, lockCtx, kvrpcpb.PessimisticLockWakeUpMode_WakeUpModeForceLock, &mutations)
+				s.NoError(err)
+
+				s.Equal(t2.GetCommitTS(), lockCtx.Values[string(key1)].LockedWithConflictTS)
+			}
+
+			// Trigger resolving locks.
+			// For normal resolve lock procedures, it always uses special path for resolving pessimistic locks, which
+			// avoids this problem. However, `BatchResolveLock` which is used by GC doesn't have that handling (and
```

Comment:
As discussed maybe we need to fix `BatchResolveLock` first and make the resolving behaviour and interface consistent

Reply:
Yes. I'll try to find some other way to do this test later.